### PR TITLE
wayland module: Allow building both client and server sides

### PR DESCRIPTION
**Pull request by xclaesse**
- Change `scope` kwarg to `public` boolean default to false.
- Change `side`  kwarg to `client` and `server` booleans.
- Document returned values
- Aggregate in a single unit test because have lots of small tests
  increases CI time.

Fixes: #10040.

### CODE REVIEW

**Comment by xclaesse** [comment]
CC @markbolhuis

**Comment by markbolhuis** [comment]
Do we want to throw an error on both client and server set to false?

**Comment by xclaesse** [comment]
> Do we want to throw an error on both client and server set to false?

That would still generate c file, is there a use case for that? I guess I should error on that, yes.

**Comment by xclaesse** [comment]
> Do we want to throw an error on both client and server set to false?

@markbolhuis Changed to make that an error. Better be more strict, we can always relax later in the unlikely event someone as a use-case I didn't think about.

**Comment by evelikov-work** [comment]
Big fan of the defaults - private + both headers. In particular:

Very few project will need public symbols. For the header side - more commonly projects will need client size only, yet the overhead of the server header is irrelevant in practise.

Small question: how does this module handle dependency of the client/server headers, where source is not needed?
Consider the following example, project X with separate internal "sub-modules", let's consider them static libraries for now:
 - sub-module A: pulls the client header, does not need the .C
 - sub-module B: pulls the server header + .C file
 - final shared library C: A + B + misc other

If I'm reading the docs correctly, to avoid multiple symbols, while ensuring proper dependency tracking one should:
 - A: declare dependency of the client header aka [1], alone
 - B: add the [0] and [2] elements as sources

Is that correct? In addition - is it possible to add symbolic names, instead of a numerical index?

**Comment by xclaesse** [comment]
@evelikov-work That's what GTK does, I made a draft port to use the module there: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/4541

tl;dr: You can split the return value to keep sources and headers separately:
```
gen = wl_mod.scan_xml(xml)
sources += gen[0]
headers += gen[1]
```

**Comment by markbolhuis** [comment]
I'd like to avoid indexing the list.
What's the point in scanning it all in one go if it's going to be split up?
If we want to keep both headers and code separate, then I suggest the following:
```
headers += wlmod.scan_headers(xmls, client : true, server : true)
src += wlmod.scan_code(xmls, public : false)
```
Thoughts?

**Comment by xclaesse** [comment]
I would e expect most of the time we don't need to split headers and source, is that more common than I thought?

**Comment by dcbaker** [comment]
I think we split them in mesa because we propagate the header, but I’d rather keep it as one call not two

**Review by dcbaker — APPROVED** [review]
